doc/sonic-application-extension/tpcm_app_ext.md

```diff
@@ -0,0 +1,467 @@
+| AEF              | Application Extension Framework       |
+
+### Document/References
+
+#### Table 2 References
+
+| **Document**                       | **Location**  |
+|------------------------------------|---------------|
+| SONiC Application Extension Infrastructure HLD | [https://github.com/stepanblyschak/SONiC/blob/8d1811f7592812bb5a2cd0b7af5023f5b6449219/doc/sonic-application-extention/sonic-application-extention-hld.md) |
+
+
+### Overview 
+
+SONiC is an open and extensible operating system. SONiC can be extended with Third Party Containers to help with Orchestration, Monitoring, and extending any features or capabilities of SONiC NOS. These can be custom built or readily available in docker community. The rich open eco system provides with multiple docker based applications which can be used in SONiC. And these need not be pre-defined packages in SONiC. While SONiC Application Extensions framework provides the infrastructure to package and integrate a TPC into SONiC, we need more capabilities to provide a dynamic environment and capabilities:
+
+- Dynamically install TPCs from various sources without pre-packaging them in SONIC. 
+- Ensure TPC has right system resources and privileges
+- Abilitiy to configure startup command arguments to TPC
+- Configure resource limits to TPCs, to ensure it does not starve core SONiC containers
+- Ability to seamlessly integrate into SONiC Services Architecture enabling start/ auto restart/ establish dependencies
+
+
+### Requirements
+
+These open TPCs help extend SONiC capabilities, and thus the following requirements are outlined to integrate them seamlessly into SONiC
+
+- Dynamically install TPC from docker registry and docker image from the local file system, SCP, SFTP, URL. These remote download options are introduced now to SONiC AEF.
+- Upgrade TPCs from docker registry and docker image from local file system, SCP, SFTP, URL
+- Provide default manifest file for docker image without manifest during install.
+- Provide runtime install capability to pass various docker startup arguments and parameters through local custom manifest file.
+- Specify system resource limits for TPCs to restrict CPU, Memory usage
+- Provide update capability to update various TPC configurations like their memory, cpu, dependencies etc. This capability is new to SONiC AEF.
+
+
+
+### Architecture Design 
+
+There is no change to the SONiC architecture. 
+
+
+### High-Level Design 
+
+SONiC Application Extension Infrastructure provides the framework to integrate SONiC compatible dockers. However there are many open source third party applications which can be installed on SONiC system, to extend the capabilities, and these typically are standalone or have less interaction with the SONiC system itself. So it is not necesary to pre define these docker applications in SONiC build, and provide their corresponding manifest.json. These TPCs can be installed dynamically on a SONiC device and can be managed.
```

Comment:
The extended version of sonic extension is the TPCM where it need not comply with sonic.

Reply:
@sg893052 So I just suggest to change the wording:
```
So it is not necesary to pre define these docker applications in SONiC build
```
to
```
So it is not necesary to for these docker applications to be SONiC compliant
```

since the first one is not true for regular (non-TPC) dockers.

Reply:
Taken care